File: extension/lib/outliers.js
// Calcul du score d'outlier pour un lot de vidéos d'une même chaîne.
// Port fidèle de l'ancien backend/app/outliers.py : chaque vidéo est comparée
// à la médiane des vues des AUTRES vidéos récentes de la même chaîne (baseline
// robuste, peu sensible aux outliers déjà présents dans l'échantillon).
(function (root) {
  function median(values) {
    const sorted = [...values].sort((a, b) => a - b);
    const mid = Math.floor(sorted.length / 2);
    return sorted.length % 2 !== 0 ? sorted[mid] : (sorted[mid - 1] + sorted[mid]) / 2;
  }

  function computeOutliers(videos, { multiplier, minViewsFloor, minHistory }) {
    const n = videos.length;
    if (n < minHistory) {
      for (const v of videos) {
        v.baselineViews = null;
        v.outlierScore = null;
        v.isOutlier = false;
      }
      return videos;
    }

    const allViews = videos.map((v) => v.viewCount);

    for (const v of videos) {
      const others = videos.reduce((acc, o, i) => {
        if (o !== v) acc.push(allViews[i]);
        return acc;
      }, []);

      const baseline = others.length >= Math.max(minHistory - 1, 3) ? median(others) : null;
      v.baselineViews = baseline;

      const score = baseline && baseline > 0 ? v.viewCount / baseline : null;
      v.outlierScore = score;
      v.isOutlier = score !== null && score >= multiplier && v.viewCount >= minViewsFloor;
    }

    return videos;
  }

  const OutlierEngine = { computeOutliers, median };
  if (typeof module !== "undefined" && module.exports) module.exports = OutlierEngine;
  else root.OutlierEngine = OutlierEngine;
})(typeof self !== "undefined" ? self : this);

File: backend/app/outliers.py
"""Calcul du score d'outlier pour un lot de vidéos d'une même chaîne.

Principe : pour chaque vidéo, on compare ses vues à la médiane des vues des
*autres* vidéos récentes de la chaîne (baseline robuste, peu sensible aux
outliers déjà présents dans l'échantillon). Si la vidéo dépasse cette
baseline d'un facteur donné ET franchit un plancher de vues absolu (pour
ignorer le bruit sur les toutes petites chaînes), elle est marquée outlier.
"""

from __future__ import annotations

import statistics
from typing import TypedDict


class VideoStats(TypedDict, total=False):
    youtube_video_id: str
    view_count: int
    baseline_views: float | None
    outlier_score: float | None
    is_outlier: bool


def compute_outliers(
    videos: list[dict],
    multiplier: float,
    min_views_floor: int,
    min_history: int,
) -> list[dict]:
    n = len(videos)
    if n < min_history:
        for v in videos:
            v["baseline_views"] = None
            v["outlier_score"] = None
            v["is_outlier"] = False
        return videos

    all_views = [v["view_count"] for v in videos]

    for v in videos:
        others = [views for vid, views in zip(videos, all_views) if vid is not v]
        if len(others) < max(min_history - 1, 3):
            baseline = None
        else:
            baseline = statistics.median(others)

        v["baseline_views"] = baseline

        if baseline and baseline > 0:
            score = v["view_count"] / baseline
        else:
            score = None

        v["outlier_score"] = score
        v["is_outlier"] = bool(
            score is not None and score >= multiplier and v["view_count"] >= min_views_floor
        )

    return videos
